lib/request-handler.js

```diff
@@ -295,71 +300,109 @@ RequestHandler.prototype.prepareOnConnection = function (queries, callbacksArray
 };
 
 /**
- * Checks if the exception is either a Cassandra response error or a socket exception to retry or failover if necessary.
+ * Checks if the error and determines if it should be ignored, retried or rethrown.
  * @param {Error} err
  * @param {Function} callback
  */
 RequestHandler.prototype.handleError = function (err, callback) {
-  //Add the error to the tried hosts
+  // add the error to the tried hosts
   this.triedHosts[this.host.address] = err;
   err['coordinator'] = this.host.address;
-  var decisionInfo = { decision: retry.RetryPolicy.retryDecision.rethrow };
-  //noinspection JSUnresolvedVariable
+  if ((err instanceof errors.ResponseError) && err.code ===  types.responseErrorCodes.unprepared) {
+    //noinspection JSUnresolvedVariable
+    return this.prepareAndRetry(err.queryId, callback);
+  }
+  var decisionInfo = this.getDecision(err);
+  if (!decisionInfo || decisionInfo.decision === retry.RetryPolicy.retryDecision.rethrow) {
+    // callback in error
+    if (this.requestOptions && this.requestOptions.captureStackTrace) {
+      utils.fixStack(this.stackContainer.stack, err);
+    }
+    if (this.request instanceof requests.QueryRequest || this.request instanceof requests.ExecuteRequest) {
+      err['query'] = this.request.query;
+    }
+    return callback(err);
+  }
+  if (decisionInfo.decision === retry.RetryPolicy.retryDecision.ignore) {
+    //Return an empty response
+    return callback(
+      null,
+      new types.ResultSet(utils.emptyObject, this.host.address, this.triedHosts, this.request.consistency));
+  }
+  return this.retry(decisionInfo.consistency, decisionInfo.useCurrentHost, callback);
+};
+
+/**
+ * @returns {{decision, useCurrentHost, consistency}}
+ */
+RequestHandler.prototype.getDecision = function (err) {
+  var requestInfo = {
+    request: this.request,
+    handler: this,
+    nbRetry: this.retryCount
+  };
+  //TODO: isServerUnhealthy is going to be called isSocketError as part of NODEJS-106 and should not call setDown()
```

Comment:
We should revisit this pull request once NODEJS-106 is merged, as it affects how socket error flag is named and the behaviour when there is a socket error.